morph/cli.py

```diff
@@ -116,6 +131,109 @@ def cmd_count(args):
         print '%d\t%s' % (c, m.show().encode('utf-8'))
 
 
+def cmd_next(args):
+    notes_path = args.notes
+    prio_path = args.prio
+    mizer = MIZERS[args.mizer]
+
+    known = load_db('known')
+
+    candidates = defaultdict(list)
+    with codecs.open(notes_path, 'r', 'utf-8') as f:
+        for line in f.readlines():
+            note = line.strip()
+            text = note.split('\t', 1)[0]
+            unknowns = [m for m in mizer.getMorphemesFromExpr(text)
+                        if m not in known.db]
+            if len(unknowns) == 1:
+                candidates[unknowns[0].show()].append(note)
+
+    with codecs.open(prio_path, 'r', 'utf-8') as f:
+        for line in f.readlines():
+            freq, m = line.strip().split('\t', 1)
+            m_cite = m.split('\t', 1)[0]
+            for cand in candidates[m]:
+                print (u'%s\t%s\t%s' % (freq, m_cite, cand)).encode('utf-8')
+
+
+def cmd_grep(args):
+    pattern_string = args.pattern
+    files = args.files
+    max_count = args.max_count
+    mizer = MIZERS[args.mizer]
+
+    pattern = parse_morpheme(pattern_string.decode('utf-8'))
+
+    count = 0
+    for path in files:
+        with codecs.open(path, 'r', 'utf-8') as f:
+            for line in f:
+                ms = mizer.getMorphemesFromExpr(line)
+                if pattern in ms:
+                    sys.stdout.write(line.encode('utf-8'))
+                    count += 1
+                    if max_count is not None and count >= max_count:
+                        return
+
+
+def clear_locs(db, pred):
+    '''Remove from `db` locations that match `pred`, and morphemes left without locations.'''
+    empty_ms = []
+    for m, locs in db.db.iteritems():
+        locs.difference_update([loc for loc in locs if pred(loc)])
+        if not locs:
+            empty_ms.append(m)
+    for m in empty_ms:
+        del db.db[m]
+
+
+def cmd_sync_known(args):
+    filenames = args.input or ['known.txt']
+    should_merge = args.merge
+
+    external_db = load_db('external', allow_missing=True)
+    if not should_merge:
+        clear_locs(external_db, lambda loc: isinstance(loc, morph.morphemes.Nowhere))
```

Comment:
Happy to use a newly-created location type instead if you prefer. I picked this one because the name seems reasonably appropriate (it's nowhere in the data otherwise available to Anki) and nothing else seems to use this type. Maybe `ExternallyKnown`? Or some other name.

I do want to give this command the semantics of "sync" rather than "add to", or at least the option of "sync". (I flip-flopped a bit in my own thinking on which should be the default.) Given the "sync" semantics, I get the ability to maintain the list of "treat these as known" morphemes in a text file that I can read and edit with any of my favorite tools, and any time I save that file just run this command to effectively save the list back into MorphMan.

So in order to do that, it's important that when syncing we remove any existing locations that were previously added by this command and aren't in the new list -- otherwise if I delete them from my text file, the deletion doesn't have any effect and the nice text file I'm looking at diverges from what MorphMan actually has on record.

Maybe a clearer way to document would be something like "Any existing morphemes in external.db which were previously inserted by this command are by default removed."? Then the name of the `Nowhere` or `ExternallyKnown` type, or whatever it might be, is an implementation detail of how we accomplish the criterion "inserted by this command".